GopherCAN.c

```diff
@@ -694,73 +700,92 @@ S8 service_can_rx_buffer(void)
+	while (message->header.DLC > 0 && message->data[message->header.DLC - 1] == 0)
 	{
-		message_to_add->dlc--;
-	}
-
-#ifdef MULTI_BUS
-	// Handle the case of the message being sent to all of the busses (ID 0)
-	if (GET_ID_DEST(message_to_add->id) == ALL_MODULES_ID)
-	{
-		send_message_to_all_busses(message_to_add);
-		return CAN_SUCCESS;
+		message->header.DLC--;
 	}
 
-	// If there are multiple busses, choose the correct bus based on the routing table
-	buffer = choose_tx_buffer_from_dest_module(message_to_add);
+	// if extended ID, get destination
+	// send standard ID data messages to all modules
+	MODULE_ID module_id = message->header.IDE == CAN_ID_EXT ?
+	        GET_ID_DEST(message->header.ExtId) :
+	        ALL_MODULES_ID;
 
-#else
-	buffer = &tx_buffer;
-#endif // MULTI_BUS
-
-	// check to make sure the buffer is not full, then add it to the back of the TX buffer
-	if (IS_FULL(buffer))
-	{
-		return TX_BUFFER_FULL;
-	}
-
-#ifdef MULTI_BUS
-	// Turn off the TX interrupt (if applicable) and add the message to the buffer
-#if TARGET == F7XX || TARGET == F4XX
-	HAL_CAN_DeactivateNotification(choose_hcan_from_tx_buffer(buffer), CAN_IT_TX_MAILBOX_EMPTY);
-#endif // TARGET == F7XX || TARGET == F4XX
-	add_message_by_highest_prio(buffer, message_to_add);
-
-#if TARGET == F7XX || TARGET == F4XX
-	HAL_CAN_ActivateNotification(choose_hcan_from_tx_buffer(buffer), CAN_IT_TX_MAILBOX_EMPTY);
-#endif // TARGET == F7XX || TARGET == F4XX
+	if (module_id == ALL_MODULES_ID) {
+#if NUM_OF_BUSSES > 2
+        add_message_by_highest_prio(&txbuff2, message);
+#endif
+#if NUM_OF_BUSSES > 1
+        add_message_by_highest_prio(&txbuff1, message);
+#endif
+        add_message_by_highest_prio(&txbuff0, message);
+    } else {
+        CAN_MSG_RING_BUFFER* buffer = choose_tx_buffer_from_dest_module(module_id);
+        add_message_by_highest_prio(buffer, message);
+    }
 
-	return CAN_SUCCESS;
+    return CAN_SUCCESS;
 }
-#else
-    add_message_by_highest_prio(buffer, message_to_add);
+
+// service_can_rx_message_std
+// handle standard ID CAN messages (data messages)
+// finds the specified group and decodes parameters
+static S8 service_can_rx_message_std(CAN_MSG* message)
+{
+    PARAM_GROUP* group = NULL;
+
+    // find the specified parameter group
+    for (U8 i = 0; i < NUM_OF_GROUPS; i++) {
+        if (GROUPS[i].id == message->header.StdId) {
+            group = &GROUPS[i];
+            break;
+        }
+    }
+
+    if (group == NULL) return NOT_FOUND_ERR;
+
+    // decode parameters
+    U8 param_start = 0;
+    U8 param_length = 1;
+    S8 err;
+    for (U8 i = 1; i < 8; i++) {
+        if (group->slots[i] == group->slots[i-1]) {
+            // count consecutive slots belonging to the same parameter
+            param_length++;
+        } else {
+            // start of a new parameter, decode previous param
+            GCAN_PARAM_ID id = group->slots[i-1];
+            if (id < EMPTY_ID || id >= NUM_OF_PARAMETERS) {
```

Comment:
We dont want to return BAD_PARAMETER_ID when there is a group that is not full

Reply:
Empty slots in a group will have EMPTY_ID, so this will only be an error if something unexpected is found in the group.

Reply:
Got it I misunderstood the code